### ISSUE TITLE
Include git hash in Agda version

### ISSUE THREAD

**Issue by Blaisorblade** [agda/agda#2023]
On a prerelease version, `agda --version` is certainly very unspecific. Other software includes additionally the git commit at build time, and a marker if the tree was dirty, and I'm already at the point where I wish it for Agda. Would you accept a PR for this?

I'm preparing a prototype based on https://mail.haskell.org/pipermail/haskell-cafe/2012-February/099654.html; the dependency is not on Stackage yet so I guess it should be added, but if it builds it should not be a problem. (Also, that's not a problem here). Additionally, this feature is not actually needed for releases.

**Comment by asr** [comment]
+1

What we need to do in releasing time for removing the Git hash from the version?

**Comment by UlfNorell** [comment]
@nad did some research on this at the latest AIM and we concluded it required more hackery than it was worth. I believe that might have been with the goal of getting the commit hash into the repository though. Just baking it in at compilation time should be less trouble.

**Comment by Blaisorblade** [comment]
> Just baking it in at compilation time should be less trouble.

Yes; using TemplateHaskell makes this simpler than going, e.g., through `Setup.hs` (see https://gist.github.com/hvr/656738, now even bitrotten).

> What we need to do in releasing time for removing the Git hash from the version?
> @asr Hopefully nothing extra — just tagging the release.

I think this forces each release to be actually tagged, but that's now part of the workflow anyway
I'm tempted to rely almost entirely on the output of `git describe`, since it's designed exactly for this purpose and it's been there since forever. When that fails we'd fallback on the current version. To ignore non-release tags, I'd restrict to tags matching `*.*.*` (that's not a regexp, but a glob pattern like on the command line).

```
$ git describe --dirty --always --match '*.*.*'
2.5.1-281-g7b863ac-dirty
```

Other questions:
- [x] agda-mode gets the version number. Will it pass the longer output through? 
- [x] agda-mode compares the version number to check compatibility of agda-mode. So, naively, developers have to reboot Emacs to reload the new Agda mode every time they install a new Agda with one more commit? That sounds bad.
  Maybe I should just keep the current string for Agda-mode consumption.
- [x] There are 3-4 files depending on Agda.Version that maybe shouldn't, and could get recompiled spuriously. Not a very big deal, but maybe I can look into it.

**Comment by andreasabel** [comment]
Ad 2: Yes, I think the agda-mode does not change so often, also not the interface to the agda-mode, so the match does not have to be commit-precise.

**Comment by asr** [comment]
> There are 3-4 files depending on Agda.Version that maybe shouldn't, and could get recompiled spuriously. Not a very big deal, but maybe I can look into it.

In releasing time, we change the version in these [files](https://github.com/agda/agda/blob/7445ce495b3caa11265571ee711cfc87aa9fb29a/notes/agda-releases-candidates#L8-L16).

Which files are you talking about?

**Comment by asr** [comment]
> > What we need to do in releasing time for removing the Git hash from the version?
> 
> @asr Hopefully nothing extra — just tagging the release.

It didn't work. After cloning your [wip/2023](https://github.com/Blaisorblade/agda/tree/wip/2023) branch and running the following commands:

```
$ git tag -a 2.5.2 -m "Agda 2.5.2 release"
$ make install-bin
```

the version reported by Agda is

```
$ agda-2.5.2 --version
Agda version 2.5.2. Git commit: 7b863ac
```

**Comment by Blaisorblade** [comment]
> > There are 3-4 files depending on Agda.Version that maybe shouldn't, and could get recompiled spuriously. Not a very big deal, but maybe I can look into it.
> > In releasing time, the files where we change the version are theses.
> 
> Which files are you talking about?

No, I simply talked about Haskell files that import `Agda.Version` hence get recompiled. I don't want to increase too much the minimal build time.

> See section "Fast builds lead to faster turnaround" on The Joy and Agony of Haskell in Production.
> 
> It didn't work.

Sorry I was unclear — I didn't implement that yet. But in that tree, `git describe --dirty --always --match '*.*.*'` should (hopefully) give 2.5.2.
BTW, since you cloned the branch: I might force-push it.

**Comment by asr** [comment]
> I might force-push it.

No yet...

Your branch didn't pass the test-suite.

**Comment by asr** [comment]
> See section "Fast builds lead to faster turnaround" on The Joy and Agony of Haskell in Production.

It is "funny" you cite this section when discussing a commit that includes TemplateHaskell.

**Comment by asr** [comment]
> Sorry I was unclear — I didn't implement that yet. But in that tree, `git describe --dirty --always --match '*.*.*'` should (hopefully) give 2.5.2.

Yep.

**Comment by andreasabel** [comment]
Nice article, the one by Stephen Diehl.

**Comment by Blaisorblade** [comment]
> Your branch didn't pass the test-suite.

That implementation is very "preliminary" — which is why I didn't even send a PR yet :-)

> It is "funny" you cite this section when discussing a commit that includes TemplateHaskell.

In case you have a technical concern: that guide discourages TemplateHaskell, indeed, but it seems this usage is the safest possible, and the cost of running `git` should hopefully dwarf the cost of interpreting the few lines to launch the process and process its result. (I'm no expert on TemplateHaskell problem, so I checked StackOverflow http://stackoverflow.com/q/10857030/53974).

**Comment by asr** [comment]
In a big project like Agda, I would like to use the `fno-code` option which (from the [documentation](http://downloads.haskell.org/~ghc/8.0.1/docs/html/users_guide/phases.html#ghc-flag--fno-code)):

> Omit code generation (and all later phases) altogether. This is useful if you’re only interested in type checking code.

This option really speeds up compilation, but I cannot use this option because it is incompatible with TemplateHaskell.

**Comment by UlfNorell** [comment]
Do you have this working now? Aren't we already using Template Haskell for stuff?

Anyway this particular use of Template Haskell could probably be ifdef'd away for `-fno-code`: just replace `gitVersion` with `Nothing` when compiling with `-fno-code`.

**Comment by asr** [comment]
> Do you have this working now?

No.

**Comment by Blaisorblade** [comment]
Right now, my bigger issue with Template Haskell is recompilation checking — `Version.hs` should be rebuilt when there's a commit. I'm tempted to add `-fforce-recomp`, but I'm trying to use `addDependentFile` to fix this properly.

Frankly, I could be fine with generating Haskell sources in a low-tech way from the `Makefile` — that would ensure recompilation (for when using the `Makefile` at least) and would avoid `TemplateHaskell` here.

For the rest, I'd personally prefer the discussion on `TemplateHaskell` in general to happen elsewhere — there are other uses which could be avoided or worked around more easily. Since I found no issue, I created #2038.

**Comment by nad** [comment]
> Frankly, I could be fine with generating Haskell sources in a low-tech way from the Makefile — that would ensure recompilation (for when using the Makefile at least) and would avoid TemplateHaskell here.

I don't think we should make it harder to build Agda. If you type either `cabal install` or `make`, then Agda should be installed (assuming that the necessary dependencies are available).

**Comment by Blaisorblade** [comment]
I was not sure how to solve all issues, but I just noticed (by chance) that `stack` has this feature and it seems to work well, based on the [gitrev](https://hackage.haskell.org/package/gitrev) package—it's a better version of `vcs-revision`, appears actually maintained, is used in `optparse-applicative`, and it seems to call `addDependentFile` often enough to track the version accurately.

As far as I can tell, their infrastructure should omit the git revision when building outside of a repo:
https://hackage.haskell.org/package/optparse-simple-0.0.3/docs/src/Options-Applicative-Simple.html#simpleVersion

Separately, they have a Cabal flag to disable git revisions in the version, but they introduced it to support developers annoyed at the extra recompilations:
https://github.com/commercialhaskell/stack/issues/1425

**Comment by gallais** [comment]
Installing `gitrev` failed on my machine using debian stable because it expects a version of `gcc` supporting the `-no-pie` command line option (whatever that is). The one on testing seems to do the trick but it seems silly to impose such a strong lower bound just for including the git hash in the version number. My `~/.cabal/logs/gitrev-1.2.0.log`:

```
Configuring gitrev-1.2.0...
Building gitrev-1.2.0...
Preprocessing library gitrev-1.2.0...
[1 of 1] Compiling Development.GitRev ( src/Development/GitRev.hs, dist-2.5.2/build/Development/GitRev.o )

<no location info>:
    Warning: Couldn't figure out linker information!
             Make sure you're using GNU ld, GNU gold or the built in OS X linker, etc.
gcc: error: unrecognized command line option ‘-no-pie’
```

**Comment by Blaisorblade** [comment]
@gallais Can you confirm `gitrev` is actually at fault, instead of "your system broke"? Reasons:
- building gitrev uses NO language extensions (it defines TemplateHaskell code, but `TemplateHaskell` is only enabled in clients)
- (EDIT) I've seen GHC breakages related to PIE on some distributions

Sources where I looked for GHC extensions (and found none):
http://hackage.haskell.org/package/gitrev-1.2.0/gitrev.cabal
http://hackage.haskell.org/package/gitrev-1.2.0/docs/src/Development-GitRev.html

**Comment by Blaisorblade** [comment]
Re the PIE-related breakage: Debian unstable/Ubuntu 16.10 enable `-fpie`, so GHC must pass `-fno-pie` to gcc/binutils. (I think https://groups.google.com/forum/#!topic/linux.debian.bugs.dist/e13FoJFKiAw is relevant, stack backported that patch to some GHC it ships).

I suspect you have a very new GHC with that fix and a very old GCC (reasonable for Debian stable) rejecting that option for some reason.

If this hypothesis is confirmed, feel free to CC me on issues you open elsewhere to debug/fix this.

**Comment by gallais** [comment]
I have been (re)compiling Agda a great many times during AIM XXIV (at least once where I had to compile all of the dependencies too) and the build only failed after I pulled the changes adding `gitrev` as a dependency.

I have `ghc-7.10.3-10` and `gcc-4:4.9.2-2` ([the one on debian stable](https://packages.debian.org/jessie/gcc)). Well, not have but had, because I've now installed `gcc-4:6.1.1-1` from testing and `gitrev` compiles just fine after the upgrade.
